rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,608 @@
+            best_merge = merge
+            best_goodness = merge.goodness
+
+    # Return the best merge and the best goodness for the calling method
+    return best_merge
+
+
+def get_all_merges(routing_table):
+    """Get possible sets of entries to merge.
+
+    Yield
+    -----
+    :py:class:`~.Merge`
+    """
+    # Memorise entries that have been considered as part of a merge
+    considered_entries = set()
+
+    for i, entry in enumerate(routing_table):
+        # If we've already considered this entry then skip
+        if i in considered_entries:
+            continue
+
+        # Construct a merge by including other routing table entries below this
+        # one which have equivalent routes.
+        merge = set([i])
+        merge.update(
+            j for j, other_entry in enumerate(routing_table[i+1:], start=i+1)
+            if entry.route == other_entry.route
+        )
+
+        # Mark all these entries as considered
+        considered_entries.update(merge)
+
+        # If the merge contains multiple entries then yield it
+        if len(merge) > 1:
+            yield Merge(routing_table, merge)
+
+
+def get_insertion_index(routing_table, generality):
+    """Determine the index in the routing table where a new entry should be
+    inserted.
+    """
+    # Wrapper for get_generality which accepts a routing entry
+    def gg(entry):
+        return get_generality(entry.key, entry.mask)
+
+    # Perform a binary search through the routing table
+    bottom = 0
+    top = len(routing_table)
+    pos = (top - bottom) // 2
+
+    pg = gg(routing_table[pos])
+    while pg != generality and bottom < pos < top:
+        if pg < generality:
+            bottom = pos  # Move up
+        else:  # pg > generality
+            top = pos  # Move down
+
+        # Compute a new position
+        pos = bottom + (top - bottom) // 2
+        pg = gg(routing_table[pos])
+
+    while (pos < len(routing_table) and
+           gg(routing_table[pos]) <= generality):
+        pos += 1
+
+    return pos
+
+
+class Merge(namedtuple("Merge", ["routing_table", "entries", "key", "mask",
+                                 "generality", "goodness",
+                                 "insertion_index"])):
+    """Represents a potential merge of routing table entries.
+    """
+    def __new__(cls, routing_table, entries=set()):
+        # Generate the new key and mask
+        any_ones = 0x00000000  # Wherever there is a 1 in *any* of the keys
+        all_ones = 0xffffffff  # ... 1 in *all* of the keys
+        all_set = 0xffffffff  # ... 1 in *all* of the masks
+
+        for i in entries:
+            # Get the entry
+            entry = routing_table[i]
+
+            # Update the values
+            any_ones |= entry.key
+            all_ones &= entry.key
+            all_set &= entry.mask
+
+        # Compute the new mask, key and generality
+        any_zeros_or_xs = ~all_ones
+        mask = all_set & (any_ones ^ any_zeros_or_xs)
```

Comment:
Could you add (yet another) intermediate step:

`changes_or_x = any_ones ^ any_zeros_or_xs`

As I understand this gives the set of bits which are either already X or which we have examples of both 1 and 0. This just makes it clearer that we're adding Xs to the mask for which we know we need to match both 0 and 1.

As an aside: the fact that there are 1s in any_zeros_or_xs and changes_or_x when we have an X really doesn't matter since all this only ends up getting ANDed with the mask in the end anyway. While I appreciate that not including the "or_x" bits of these names is incorrect (and possibly going to worry a careful reader), including them will also worry a potential reader looking for why it matters (see: me!). I'm not sure whether this justifies a short comment to reassure or not...